Improve efficiency of QkCircuitInstruction in C API

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adjusts the decleration and usage of the QkCircuitInstruction struct in the circuits C API. The larger change is around the interface for the `qk_circuit_get_instruction` function which has changed from returning a struct by value to writing it to a user provided pointer. From a strictly ergonomics PoV the normal use case of `qk_circuit_get_instruction()` function will typically be used while iterating over the circuit or getting the details of multiple instructions, so being able to reuse a single allocation is quite useful.



### Details and comments

Fixes #14390

mtreinish committed:
Improve efficiency of QkCircuitInstruction in C API

This commit adjusts the decleration and usage of the
QkCircuitInstruction struct in the circuits C API. The larger change is
around the interface for the `qk_circuit_get_instruction` function which
has changed from returning a struct by value to writing it to a
user provided pointer. From a strictly ergonomics PoV the normal use
case of `qk_circuit_get_instruction()` function will typically be used
while iterating over the circuit or getting the details of multiple
instructions, so being able to reuse a single allocation is quite
useful.

Fixes #14390

mtreinish committed:
Merge branch 'main' into update-c-instruction-api

mtreinish committed:
Merge remote-tracking branch 'origin/main' into update-c-instruction-api

mtreinish commented on crates/cext/src/circuit.rs:
```diff
@@ -514,11 +514,14 @@ pub struct CInstruction {
+///
 ///
 /// @return The instruction details for the specified instructions.
 ///
```
```suggestion
```

mtreinish committed:
Update crates/cext/src/circuit.rs

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -514,11 +514,12 @@ pub struct CInstruction {
 ///
 /// @param circuit A pointer to the circuit to get the instruction details for.
 /// @param index The instruction index to get the instruction details of.
+/// @param instruction A pointer to where to write out the ``QKCircuitInstruction``
```
```suggestion
/// @param instruction A pointer to where to write out the ``QkCircuitInstruction``
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -492,18 +492,18 @@ pub unsafe extern "C" fn qk_circuit_num_instructions(circuit: *const CircuitData
+    /// The number of clbits for this instruction.
+    num_clbits: u32,
+    /// The number of parameters for this instruction.
+    num_params: u32,
```
Looks good, I checked the size decreases from 56 bits to 48 👍🏻 I'm mildly unhappy when thinking about initializing this from C side, which would now be 
```
QkCircuitInstruction inst = {name, qubits, clbits, params, num_qubits, num_clbits, num_params};
```
instead of having `pointer, pointer_length` grouped -- but speed > style 🤷🏻

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
-        clbits: out_cargs,
-        num_params: params.len() as u32,
-        params: out_params,
+    unsafe {
```
Safety police: This needs a Safety doc -- alternatively it could be done by casting the pointer into a reference and writing the values as in 
https://github.com/Qiskit/qiskit/blob/308df7ac22ff91d9f49fbdb6998578dbcda60ba9/crates/cext/src/sparse_observable.rs#L294

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/14404/commits/1b8943aa13df616319b1f53b36bef07cc9b13623

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
-        num_params: params.len() as u32,
-        params: out_params,
+    unsafe {
+        *instruction = CInstruction {
```
Follow-up question: why is Rust fine instantiating the temporary `CInstruction` object here? I would've expected cargo to complain that `instruction` is set to a value that goes out of scope, given that `CInstruction` doesn't implement `Copy` -- but I guess it just happily moves ownership here somehow?

mtreinish replied:
It's moving the data to the pointer that's what the deref is doing here. I'll write this out as:

```rust
unsafe {
    std::ptr::write(instruction, CInstruction {...})
}
```

I think this makes it a bit more clear although it's essentially the same thing although I think it's probably a bit safer.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14404/commits/1b8943aa13df616319b1f53b36bef07cc9b13623

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
+/// responsible for allocating and freeing the actually allocation used to store a
+/// ``QkCircuitInstruction``.
+///
 /// @param inst The instruction to free.
```
```suggestion
/// @param inst A pointer to the instruction to free.
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
 /// @ingroup QkCircuit
 /// Free a circuit instruction object.
 ///
+/// This function doesn't free the allocation for the provided ``QkInstruction`` pointer, it
```
```suggestion
/// This function doesn't free the allocation for the provided ``QkCircuitInstruction`` pointer, it
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
 ///
+/// This function doesn't free the allocation for the provided ``QkInstruction`` pointer, it
+/// only free the internal allocations for the data contained in the instruction. You are
+/// responsible for allocating and freeing the actually allocation used to store a
```
```suggestion
/// responsible for allocating and freeing the actual allocation used to store a
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
 /// @ingroup QkCircuit
 /// Free a circuit instruction object.
 ///
+/// This function doesn't free the allocation for the provided ``QkInstruction`` pointer, it
```
So this is a bit different to `qk_obs_term`, which also writes the observable term into a pre-allocated variable: the `QkObsTerm` only points to existing data living in a `QkObs`, whereas here we copy the qubit/clbits/params and they can live outside the `QkCircuit` that they came from.

Do I then correctly understand the flow as:
```c
QkCircuit *circuit = // some circuit coming from somewhere
QkCircuitInstruction *inst = malloc(sizeof(QkCircuitInstruction));
for (size_t i = 0; i < qk_circuit_len(circuit); i++) {
   qk_circuit_get_instruction(circuit, i, inst);
   // do something with the instruction...
   qk_circuit_instruction_free(inst);  // free the data inside the instruction
}
free(inst);  // free the memory allocation of the instruction itself
qk_circuit_free(circuit);  // free the circuit
```

It doesn't quite fit the current flow where `qk_<object>_free` actually frees the object itself. I wonder how this is dealt with in other situations, if there's the norm of having something like `qk_<object>_free` that frees both internal data + object, plus `qk_<object>_clear` or `qk_<object>_free_data` or sth to free the internal data. This is not something to change, just to brainstorm what users would expect 🙂

mtreinish replied:
yeah that's the basic idea. The `qk_circuit_instruction_free` function frees the internal rust managed allocations that are pointed to by the `CircuitInstruction` but the user manages the allocation for the struct itself. I can rename the method clear if you think that makes it more obvious.

mtreinish replied:
I renamed it `qk_circuit_instruction_clear()` to address this: https://github.com/Qiskit/qiskit/pull/14404/commits/41b7061d9e02212c63d5f16abe0683c88dc9aab1 I think this should address the concern here. I agree using the `free` suffix here is going against our naming convention for the rest of the library.

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
+///     qk_circuit_gate(qc, QkGate_H, *[0], *[]);
+///     qk_circuit_get_instruction(qc, 0, inst);
+///     qk_circuit_free(inst);
+///     free(inst)
```
```suggestion
///     qk_circuit_instruction_free(inst); // free the data
///     free(inst); // free the pointer
///     qk_circuit_free(qc); // free the circuit
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -560,31 +563,48 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
+///
+///     QkCircuitInstruction *inst = malloc(sizeof(QkCircuitInstruction));
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_gate(qc, QkGate_H, *[0], *[]);
```
Error: Wishful syntax 🙂

mtreinish replied:
More just sloppy copy and paste from someone's suggestions on #14006 

```suggestion
///     qk_circuit_gate(qc, QkGate_H, *[0], NULL);
```

Cryoris reviewed (commented):
* The data reshuffling LGTM 👍🏻 
* I'm also a fan of populating `QkCircuitInstruction` by pointer, just like `qk_obs_term`
* Below I have some questions on the bullet above 🙂

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Update crates/cext/src/circuit.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Merge remote-tracking branch 'origin/main' into update-c-instruction-api

mtreinish committed:
Use std::ptr::write instead of *ptr =

mtreinish committed:
Update unitary gate tests to use new get_instruction syntax

mtreinish committed:
Rename qk_circuit_instruction_free to qk_circuit_instruction_clear

This commit renames the qk_circuit_instruction_free() function to
qk_circuit_instruction_clear(). The function isn't freeing the
allocation for the QkCircuitInstruction object but instead is clearing
out internal allocations. This rename makes it clear that the behavior
here is eparate from the convention on what we're using for qk_*_free()
functions.

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -874,31 +877,54 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
+///
+///     QkCircuitInstruction *inst = malloc(sizeof(QkCircuitInstruction));
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_gate(qc, QkGate_H, *[0], NULL);
```
some habits die hard 😉 
```suggestion
///     uint32_t q0 = {0};
///     qk_circuit_gate(qc, QkGate_H, q0, NULL);
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -874,31 +877,54 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
 ///
 /// # Safety
 ///
 /// Behavior is undefined if ``inst`` is not an object returned by ``qk_circuit_get_instruction``.
```
This should probably be something like this now?
```suggestion
/// Behavior is undefined if ``inst`` is not a valid, non-null pointer to a ``QkCircuitInstruction``.
```

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Merge branch 'main' into update-c-instruction-api

Cryoris reviewed (approved):
LGTM thanks for the updates!
